### DISCUSSION TITLE
Multiple structure instances in a Molstar viewer

### DISCUSSION THREAD

**Discussion by 0gust1** [molstar/molstar#475] [Q&A]
Hi, 

We're currently working to replace NGL with Mostar in our web application.

One thing that we achieved to do with NGL, is to display the same structure several times, with a different residue coloring scheme.

This kind of view is **very** useful to compare the results of several different dockings or epitope mappings, see the screenshot below.

![Capture d’écran 2022-07-05 à 18 05 16](https://user-images.githubusercontent.com/578154/177371084-7ae06054-1aff-48cc-b679-a262000b6855.png)

In the example above:
1. rotation / zoom of each structure is synchronized with the others
2. We overlay some info for each structure (here, name and bin number), as HTML over the NGL canvas 
3. it works well but is not very optimized (we add 9 times the same structure to the 3D scene, and the surface is also computed/rendered 9 times... leading to crashes with bigger batches and big/complicated proteins).
4. As you can see, each structure receive its own residue coloring scheme

**Questions:**

How could we approach this with Molstar ? Especially with the points 1 and 3 above (the point 3 being a bit a pain).   
_edit:_ is this what the brand new param/feature `instanceGranularity` is about ? if yes, can someone explain me roughly the concept ?

I take anything: ideas/directions, pointers to specific Molstar modules/examples, pseudocode, etc. It will anyway help me becoming familiar enough with the tool. :-)

**Comment by dsehnal** [comment]
This is possible with Mol* too, here is an example from [Envision](https://www.entos.ai/envision):

https://user-images.githubusercontent.com/21145776/177492176-c65a50b4-8f99-4d16-9252-bde1a4c4fc64.mp4

----------

The simplest way to do this would probably be:

- You create multiple plugin contexts and load the structure in each of them (state sharing is something we have planned) and apply the coloring you want.
- Set ``Canvas3DParams.viewport`` for each frame to ``relative-frame`` with appropriate coords.
- Subscribe to ``plugin.canvas3d.didDraw`` for each context and update all the other contexts with the current camera position of the one you are interacting with.
  - You need to create an overlay over the plugins to determine which one is active and should be used as source for camera position.
  - Setting ``Canvas3DParams.camera.manualReset`` can help.

-------------

Alternatively, you can skip the ``PluginContext`` approach and use the ``Canvas3D`` (with ``viewport`` param will strill work) directly. With this approach you can share the state, but would have to re-implement some logic the ``PluginContext`` provides. To see an example of this you can have a look in ``src/tests/browser``.

**Comment by 0gust1** [comment]
Hi @dsehnal , I had some time to dig further with Molstar + svelte integration. I have more questions...

> Set Canvas3DParams.viewport for each frame to relative-frame with appropriate coords.

I don't understand what `relative-frame` do (and why I should use it).  Could you elaborate quickly ?

> - Subscribe to plugin.canvas3d.didDraw for each context and update all the other contexts with the current camera position of the one you are interacting with.
>   - You need to create an overlay over the plugins to determine which one is active and should be used as source for camera position.
>   - Setting Canvas3DParams.camera.manualReset can help

`didDraw` Observable is also updated for hover interactions on residues/AA. How can I discriminate between hover-triggered `didDraw` and rotation/zoom-triggered `didDraw` ? 
=> Hover-triggered ones should update the highlight and rotation/zoom-triggered ones should update the camera (update = update on all plugins except the one that is "active"). 

As Svelte has a native support of Observables (with Svelte stores), I plan to leverage Molstar's RxJs/signals/BehaviorSubject as possible.

**Reply by dsehnal** [reply]
Relative frame says how much of the canvas should the viewport take when sharing a single webgl context. It's similar to calling glViewport function (hope I got the name right)

---------

If you want unique camera positions, subscribe to the didDraw observable which you map to a `canvas.camera.snapshot()` and pipe it through `disctintUntilChanged` to get a sequence of unique camera positions. A similar approach is already used in volume streaming behavior in the source code I think.

**Reply by 0gust1** [reply]
Damn, your comment made me realize I'm not doing it the right way: I'm currently instanciating several pluginContext, but each one with its own canvas... 

In the documentation it is stated:

> A single-plugin setup is shown the example below. In order to initialize multiple plugins, each with its own context and viewport, some extra steps are required (docs section to be added).

Where can I find an example of multiple viewports on a single canvas?  I'm really sorry if it sounds a dumb question.

**Reply by dsehnal** [reply]
There isn't such example -- that's what "later" means sometimes :)

What you need to do is to create `Canvas3DContext.fromCanvas` object manually. You can then pass this object to `PluginContext.initContainer/mount/initViewer/Canvas3D.create` depending on how you plan to use the library.

Then, for example, a relative frame `{ x: 0.25, y: 0.25, width: 0.25, height: 0.25 }` for one your canvases will use this subsection of the viewport.

------------

Also, it would probably be fine to use separate webgl contexts if the number of grid cell you need isn't too high.

**Reply by dsehnal** [reply]
One more detail for the unique camera positions is that you need to use `Camera.areSnapshotsEqual` to compare the values inside the `disctintUntilChanged`.

**Comment by rjdirisio** [comment]
Hi @dsehnal , I was wondering if you had specific guidance for these instructions:

> - You create multiple plugin contexts and load the structure in each of them (state sharing is something we have planned) and apply the coloring you want.
> - Set Canvas3DParams.viewport for each frame to relative-frame with appropriate coords.
> - Subscribe to plugin.canvas3d.didDraw for each context and update all the other contexts with the current camera position of the one you are interacting with.
>    - You need to create an overlay over the plugins to determine which one is active and should be used as source for camera position.
>    - Setting Canvas3DParams.camera.manualReset can help.

I have multiple plugin contexts, and I have a listener for `plugin.canvas3d.didDraw`. I am able to get the position via `plugin.canvas3d.camera.position` (is that right?). Then, what do I pass to the other plugins to update the position? I noticed there's a `plugin?.canvas3d?.camera.center`, but I don't think that's what I'm looking for.

Additionally, in my instance, I want to avoid updating the positions when `didDraw` gets triggered when the size of the window/viewport changes. Is there any simple way to do that?

Thanks!

**Reply by dsehnal** [reply]
You should update the camera positions using `get/setSnapshot` functions. Should also set the transition duration to 0 to avoid the camera animation.

I can't think of a simple way to determine didDraw was called as a result of a resize. Can you be more specific about the use case you are after?

EDIT: actually, you can do that. The way i solved this in the last was to trigger the sync based on the cursor position: because you need to determine which if the viewports is the pivot when interacting.

**Reply by rjdirisio** [reply]
Thanks, really appreciate the quick response! I implemented a version of this and it's working pretty well. There's a little bit of lag because of the subscribing / updating paradigm. 

I have two follow ups:
1. Is there still a 'single canvas' style that one can implement? This was written 3 years ago so I was wondering if the canvas implementation is still possible.
2. Alternatively, is there a way to simply share a Camera between plugins?

**Reply by dsehnal** [reply]
> Is there still a 'single canvas' style that one can implement? This was written 3 years ago so I was wondering if the canvas implementation is still possible.

What do you mean? You can use a single canvas with a single WebGL context to render multiple views.

> Alternatively, is there a way to simply share a Camera between plugins?

I suppose you could implement a custom overlay + "MultiViewportTrackball" that would communicate the camera position to several plugins at the same time. This would eliminate the lag.

**Reply by rjdirisio** [reply]
> What do you mean? You can use a single canvas with a single WebGL context to render multiple views.

Got it, just wanted to make sure that was still possible.

> I suppose you could implement a custom overlay + "MultiViewportTrackball" that would communicate the camera position to several plugins at the same time. This would eliminate the lag.

Interesting, I may go this way eventually, but I think the method described above is best for me right now. Thanks again for the quick responses
